### ISSUE TITLE
NEVER_BLOCK flag for a Request to prevent deadlock if ItemPipeline Deferred results in new requests

### ISSUE THREAD

**Issue by djay** [scrapy/scrapy#6407]
## Summary

The ability to specify a additional level of priority for a request using a flag for when you are creating requests that could cause deadlocks. For example when requests come from an ItemPipeline which returns a Deferred that relies on more requests being processed.

## Motivation

Follows on from https://github.com/scrapy/scrapy/issues/6206

Let's say ItemPipeline that is designed to add extra information to the item. And it does this by doing more requests.

The slot still considers this an active request. This is because the response is still being kept in memory.
The response isn't available in the ItemPipeline. It has been kept around mainly for logging if there is a failure.

https://github.com/scrapy/scrapy/blob/master/scrapy/core/scraper.py#L322

https://github.com/scrapy/scrapy/blob/master/scrapy/core/scraper.py#L375

As soon Deferred items responses add up to SCRAPER_SLOT_MAX_ACTIVE_SIZE, the slot will stop processing requests and there is now no way for the requests that would unlock the deferred from getting processed so it's a deadlock.

There doesn't seem to be a way to prevent this. You can't set SCRAPER_SLOT_MAX_ACTIVE_SIZE big enough.
and prioritisation doesn't prevent the situation where you could have already reached your max before the ItemPipeline has kicjed off the addition requests.
If you are aware of the possibility of the deadlock you could use this priority feature to ensure these requests don't end up creating a deadlock. they could processed one at a time at least so as not to max out the memory.

## Describe alternatives you've considered

Another way to solve it would be to delete the response in the case where an ItemPipeline exists and a deferred has been returned. A dummy response could be put in it's place to would blow up if any logging actually wanted to use the response. This would minimise the RAM usage. If they did try do something with the response there would be an error message telling the user to use a setting to override this behaviour if they are sure deadlocks are not possible

Maybe some kind of queue system where there is a way to reserve some capacity for kinds of requests whereas normal requests are limited to lower throughput than the total?

## Additional context

Follows on from https://github.com/scrapy/scrapy/issues/6206 which is about how its obvious a deadlock is occurring or why.

**Comment by Gallaecio** [comment]
If your code looks like the one you shared [here](https://github.com/scrapy/scrapy/issues/6206#issuecomment-1894775709), I think what you need to do is change your implementation. It is not clear to me where those deferreds come from, but I wonder if you would have an issue if instead of creating Deferreds elsewhere (from requests yielded from callbacks?) you used `crawler.engine.download` from the item pipeline itself to run desired requests, bypassing the request scheduler.

Also, I wonder why you are relying on the item pipeline to handle these requests in the first place. Usually, if you need an item to have data from multiple requests, you chain those requests, i.e. request 1 yields request 2 with the item data collected so far in cb_kwargs/meta, request 2 yields request 3 the same way, and so on until all requests have been processed, and the last request yields the final item with all the data.

**Comment by djay** [comment]
What I was tryng to do is make middleware that I can add into any crawler that would go and get information (change history in this case) about a given request. I was setting special objects into the item on where the information should be put, and setting metadata to tell the middleware to fill that information in for this particular request.

Using  ```crawler.engine.download``` might solve it but because its actually many requests, many of which can be done in parallel, thats going to make it slow.

Maybe you are right, maybe reverting it back chaining the requests is the way to go. I was just looking for a more plugin middleware way to do the same thing. 
I do wonder however, what the purpose of returning Deferred is from process_item if not to wait for more data. Maybe there needs to be a warning in the docs to never add more requests when using Deferred as this can cause a deadlock?

**Comment by djay** [comment]
@Gallaecio sorry, had another look at the code. It currently does use ```crawler.engine.download```. The scheduler isn't part of the deadlock problem. It's a the slot backoff due to the max active size and that must still hold with requests directly to crawler.engine.download

**Comment by Gallaecio** [comment]
Are you sure requests handled with `crawler.engine.download` count towards the max active size? I was under the impression that those did not count, since they do not go through item pipelines or anything like that.

**Comment by djay** [comment]
@Gallaecio yes it certainly seems to be. 
Otherwise I can't explain why its getting deadlocked and why increasing the max size delays the deadlock.

But I took your advice and rewrote it. How it is just a SpiderMiddlware and the original request will ask the middleware to return the item onces it's finished. It isn't yet as flexible where the item could be returned by one of multiple middlewares that need to parse more urls to add more information but it could be made to do this.

As for this enhancement request. I think at the very least it would be good to put in a warning in the ItemPipeline documentation that Deferred shouldn't be used when it results in waiting on additional requests as this can cause a deadlock. 
Some kind of logging that makes it obvious that a deadlock is happening  (https://github.com/scrapy/scrapy/issues/6206) might certainly help someone in the future waste less time.

**Comment by Gallaecio** [comment]
If we can build a minimal reproducible example that shows how to reproduce the deadlock, I agree on the docs warning, and maybe looking further into the issue, in case there is a bug there, although it might be intended behavior.
